src/types/function/mod.rs

```diff
@@ -0,0 +1,217 @@
+//! Types and traits for working with JavaScript functions.
+
+use crate::context::Context;
+use crate::handle::Handle;
+use crate::result::{JsResult, NeonResult};
+use crate::types::{JsFunction, JsObject, JsValue, Value};
+
+use smallvec::smallvec;
+
+pub(crate) mod private;
+
+/// A builder for making a JavaScript function call like `parseInt("42")`.
+///
+/// The builder methods make it convenient to assemble the call from parts:
+/// ```
+/// # use neon::prelude::*;
+/// # fn foo(mut cx: FunctionContext) -> JsResult<JsNumber> {
+/// # let global = cx.global();
+/// # let parse_int = global.get(&mut cx, "parseInt")?;
+/// # let parse_int: Handle<JsFunction> = parse_int.downcast_or_throw(&mut cx)?;
+/// let x: Handle<JsNumber> = parse_int
+///     .call_with(&cx)
+///     .arg(cx.string("42"))
+///     .apply(&mut cx)?;
+/// # Ok(x)
+/// # }
+/// ```
+#[derive(Clone)]
+pub struct CallOptions<'a> {
+    pub(crate) callee: Handle<'a, JsFunction>,
+    pub(crate) this: Option<Handle<'a, JsValue>>,
+    pub(crate) args: private::ArgsVec<'a>,
+}
+
+impl<'a> CallOptions<'a> {
+    /// Set the value of `this` for the function call.
+    pub fn this<V: Value>(&mut self, this: Handle<'a, V>) -> &mut Self {
+        self.this = Some(this.upcast());
+        self
+    }
+
+    /// Add an argument to the arguments list.
+    pub fn arg<V: Value>(&mut self, arg: Handle<'a, V>) -> &mut Self {
+        self.args.push(arg.upcast());
+        self
+    }
+
+    /// Replaces the arguments list with the given arguments.
+    pub fn args<A: Arguments<'a>>(&mut self, args: A) -> &mut Self {
+        self.args = args.into_args_vec();
+        self
+    }
+
+    /// Make the function call. If the function returns without throwing, the result value
+    /// is downcast to the type `V`, throwing a `TypeError` if the downcast fails.
+    pub fn apply<'b: 'a, V: Value, C: Context<'b>>(&self, cx: &mut C) -> JsResult<'b, V> {
+        let this = self.this.unwrap_or_else(|| cx.null().upcast());
```

Comment:
I think this should be `undefined` instead of `null`. We should also probably have a test that the correct `this` is being set; I don't see one.

Reply:
I agree, I wonder why I had the instinct to choose `null`. Here's I think a strong argument in favor of `undefined`:

```
% node
Welcome to Node.js v16.8.0.
Type ".help" for more information.
> function strict() { 'use strict'; return this }
undefined
> strict()
undefined
```